compiler/internal/dce/filters.go

```diff
@@ -2,31 +2,339 @@ package dce
+	parts := make([]string, len(tArgs))
+	for i, tArg := range tArgs {
+		parts[i] = gen.Type(tArg)
+	}
+	return `[` + strings.Join(parts, `, `) + `]`
+}
+
+// Tuple returns the filter part containing parameter or result
+// types for a function, e.g. `(int,string)`, `(int,...string)`.
+func (gen *filterGen) Tuple(t *types.Tuple, variadic bool) string {
+	count := t.Len()
+	parts := make([]string, count)
+	for i := range parts {
+		argType := t.At(i).Type()
+		if i == count-1 && variadic {
+			if slice, ok := argType.(*types.Slice); ok {
+				argType = slice.Elem()
+			}
+			parts[i] = `...` + gen.Type(argType)
+		} else {
+			parts[i] = gen.Type(argType)
+		}
+	}
+	return strings.Join(parts, `, `)
+}
+
+// Type returns the filter part for a single type.
+func (gen *filterGen) Type(typ types.Type) string {
+	switch t := typ.(type) {
+	case types.Object:
+		return gen.Object(t, nil)
+
+	case *types.Array:
+		return `[` + strconv.FormatInt(t.Len(), 10) + `]` + gen.Type(t.Elem())
+	case *types.Chan:
+		return `chan ` + gen.Type(t.Elem())
+	case *types.Interface:
+		return gen.Interface(t)
+	case *types.Map:
+		return `map[` + gen.Type(t.Key()) + `]` + gen.Type(t.Elem())
+	case *types.Named:
+		// Get type args from named instance not generic object
+		return gen.Object(t.Obj(), getTypeArgs(t))
+	case *types.Pointer:
+		return `*` + gen.Type(t.Elem())
+	case *types.Signature:
+		return `func` + gen.Signature(t)
+	case *types.Slice:
+		return `[]` + gen.Type(t.Elem())
+	case *types.Struct:
+		return gen.Struct(t)
+	case *types.TypeParam:
+		return gen.TypeParam(t)
+	default:
+		// Anything else, like basics, just stringify normally.
+		return t.String()
+	}
+}
+
+// Union returns the filter part for a union of types from an type parameter
+// constraint, e.g. `~string|int|~float64`.
+func (gen *filterGen) Union(u *types.Union) string {
+	parts := make([]string, u.Len())
+	for i := range parts {
+		term := u.Term(i)
+		part := gen.Type(term.Type())
+		if term.Tilde() {
+			part = "~" + part
+		}
+		parts[i] = part
+	}
+	// Sort the union so that "string|int" matches "int|string".
+	sort.Strings(parts)
+	return strings.Join(parts, `|`)
+}
+
+// Interface returns the filter part for an interface type or
+// an interface for a type parameter constraint.
+func (gen *filterGen) Interface(inter *types.Interface) string {
+	// Collect all method constraints with method names and signatures.
+	parts := make([]string, inter.NumMethods())
+	for i := range parts {
+		fn := inter.Method(i)
+		parts[i] = fn.Id() + gen.Signature(fn.Type().(*types.Signature))
+	}
+	// Add any union constraints.
+	for i := 0; i < inter.NumEmbeddeds(); i++ {
+		if union, ok := inter.EmbeddedType(i).(*types.Union); ok {
+			parts = append(parts, gen.Union(union))
+		}
+	}
+	// Sort the parts of the interface since the order doesn't matter.
+	// e.g. `interface { a(); b() }` is the same as `interface { b(); a() }`.
+	sort.Strings(parts)
+
+	if len(parts) <= 0 {
```

Comment:
nit/optional
```suggestion
	if len(parts) == 0 {
```

Reply:
😆 That's muscle memory from working in C/C++ for so many years (same with me having to fix any `++i` before it compiles). I fixed all the cases like that.